presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java

```diff
@@ -8247,6 +8247,53 @@ public void testRightJoinWithEmptyProbeSide()
                 "WITH small_part AS (SELECT * FROM part WHERE name = 'a') SELECT lineitem.orderkey FROM small_part RIGHT JOIN lineitem ON  small_part.partkey = lineitem.partkey");
     }
 
+    @Test
+    public void CardinalityOfFilterSimplification()
+    {
+        MaterializedResult result = computeActual("" +
+                "SELECT cardinality(filter(array_agg(orderkey), x -> x % 2 = 0)) " +
+                "FROM orders " +
+                "WHERE orderkey < 20");
+
+        assertEquals(result.getRowCount(), 1);
+        assertEquals(result.getMaterializedRows().get(0).getField(0), 3L);
+    }
+
+    @Test
+    public void CardinalityOfFilterComparisonSimplification()
```

Comment:
If we have optimized `cardinality(filter(array_agg(orderkey), x -> x < 3))`, would optimizing `cardinality(filter(array_agg(orderkey), x -> x < 3)) = 2` be an overkill?  `=2` seems to be very low cost operations. 

Also, it might make sense to add some benchmark similar to `BenchmarkArrayFilter`, or even do a comparison run from Tpch query runner and add the result as part of commit message.

Reply:
The main thing here is to avoid creating the intermediate array.

Reply:
> If we have optimized `cardinality(filter(array_agg(orderkey), x -> x < 3))`, would optimizing `cardinality(filter(array_agg(orderkey), x -> x < 3)) = 2` be an overkill? `=2` seems to be very low cost operations.
> 
> Also, it might make sense to add some benchmark similar to `BenchmarkArrayFilter`, or even do a comparison run from Tpch query runner and add the result as part of commit message.

The short-circuiting helps when you have a lot of medium-size arrays